File: src/com/saucelabs/sauceconnect/Timeout.java
package com.saucelabs.sauceconnect;

public abstract class Timeout extends Thread {
    private long millis;
    
    private volatile boolean success = false;
    private volatile boolean error = false;
    
    private Object monitor = new Object();
    
    public Timeout(long millis){
        this.millis = millis;
    }
    
    public void go() {
        start();
        try {
            synchronized(monitor){
                long start = System.currentTimeMillis();
                monitor.wait(millis);
                if(System.currentTimeMillis()-start >= (millis)){
                    if(this.isAlive()){
                        this.interrupt();
                    }
                    success = false;
                }
            }
        } catch(InterruptedException e) {
            success = false;
        }
    }
    
    public void run() {
        try {
            longRunningTask();
            success = true;
        } catch (InterruptedException e){
            success = false;
        } catch (Exception e) {
            error = true;
        } finally {
            synchronized(monitor){
                monitor.notify();
            }
        }
    }
    
    public boolean isSuccess() {
        return success;
    }

    public boolean isError() {
        return error;
    }

    public abstract void longRunningTask() throws Exception;
}

File: src/com/saucelabs/sauceconnect/Timeout.scala
package com.saucelabs.sauceconnect

abstract class Timeout(millis:Long) extends Thread {
    @volatile
    private var success = false
    @volatile
    private var error = false

    private var monitor = new Object()

    def go() = {
        start()
        try {
            monitor.synchronized{
                val start = System.currentTimeMillis()
                monitor.wait(millis)
                if(System.currentTimeMillis()-start >= (millis)){
                    if(this.isAlive()){
                        this.interrupt()
                    }
                    success = false
                }
            }
        } catch {
            case e:InterruptedException => success = false
        }
    }

    override def run() = {
        try {
            longRunningTask()
            success = true
        } catch {
            case e:InterruptedException => success = false
            case e => error = true
        } finally {
            monitor.synchronized{
                monitor.notify()
            }
        }
    }

    def isSuccess() = success
    def isError() = error

    @throws(classOf[Exception])
    def longRunningTask()
}
